### PR TITLE
feat(core): Add storage upload to move away from unified upload protocol

### PR DESCRIPTION
**Pull request by bajajneha27**

### CODE REVIEW

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:55]: Why is this necessary?

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:97]: I assume the idempotency considerations for the retry are handled at the level of the manual storage client?

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:108]: Is there a standard for what gets logged in the Ruby client? Just curious since we don't really do debug logging in Go currently.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:144]: What happens if a chunk upload fails and the input has already been read? Is the upload_io able to rewind? I don't really see how this happens in the code (but my ruby isn't very good!).

File: google-apis-generator/lib/google/apis/generator/templates/_method.tmpl
Comment (tritone) [inline google-apis-generator/lib/google/apis/generator/templates/_method.tmpl:57]: Does this need to be changed in the generated code in this PR also? In Go, presubmits will fail if the generated code doesn't match.

**Review by tritone — COMMENTED** [review]
Generally looking good, nice progress here! A few things, mostly questions.

File: google-apis-generator/lib/google/apis/generator/templates/_method.tmpl
Comment (bajajneha27) [reply google-apis-generator/lib/google/apis/generator/templates/_method.tmpl:57]: Once this change is merged, the generator will generate the new code and raise a PR. Or I'll do it manually.

**Comment by bajajneha27** [comment]
@tritone Integration tests are already there in [google-cloud-ruby](https://github.com/googleapis/google-cloud-ruby/blob/main/google-cloud-storage/acceptance/storage/file_test.rb). I tested them against this code change and they ran successfully. The tests look thorough enough to me and covers all cases. So these tests along with conformance tests should be enough for our testing, I believe?

File: google-apis-core/spec/google/apis/core/storage_upload_spec.rb
Comment (tritone) [reply google-apis-core/spec/google/apis/core/storage_upload_spec.rb:19]: I think as long as the `google-cloud-ruby` integration tests pass with these changes, we are good.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [reply google-apis-core/lib/google/apis/core/storage_upload.rb:110]: Ah I wasn't sure how you decide which things to log using debug logging. Why here and not other places? Although maybe this is just copied from existing code?

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (tritone) [reply google-apis-core/lib/google/apis/core/storage_upload.rb:147]: Okay, gotcha -- so the rewinding happens in L153 I see

**Review by tritone — APPROVED** [review]
LGTM, pending approval from a Ruby reviewer and confirmation that the google-cloud-ruby storage integration tests pass.

File: google-apis-core/lib/google/apis/core/base_service.rb
Comment (dazuma) [inline google-apis-core/lib/google/apis/core/base_service.rb:326]: `Symbol` is a class and should be capitalized

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (dazuma) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:30]: I know we're not running rubocop on this repo yet, but let's start using our style for new code (e.g. double quotes for all strings, omit parens in method calls further down, etc.)

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (dazuma) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:69]: I think `type&.content_type` is a bit more clear.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (dazuma) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:162]: Fixnum is defunct as of Ruby 2.4. Use Integer.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (frankyn) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:73]: It would help to know which type is not supported.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (frankyn) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:76]: GCS handles the default value for users; `X-Goog-Upload-Content-Type` is optional unless that's changed recently.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (frankyn) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:145]: IIUC, this code accepts a streamable type such as `STDIN` which may know the `upload_io.size`.

You can fall back to unknown sized resumable uploads: 
image.

source: https://cloud.google.com/storage/docs/streaming#rest-apis

File: google-apis-core/spec/google/apis/core/storage_upload_spec.rb
Comment (frankyn) [inline google-apis-core/spec/google/apis/core/storage_upload_spec.rb:25]: Please include IO type which does not have a known size in your tests.

**Review by frankyn — CHANGES_REQUESTED** [review]

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (bajajneha27) [reply google-apis-core/lib/google/apis/core/storage_upload.rb:145]: Discussed offline. We'll implement this in a separate PR. Created b/246693802 to track it.

File: google-apis-core/lib/google/apis/core/storage_upload.rb
Comment (dazuma) [inline google-apis-core/lib/google/apis/core/storage_upload.rb:146]: Small nit here, but I'd name the `body` local variable something like `chunk_body` instead, to distinguish it from `self.body` which is a method on the base class (and is treated as such in line 120). This will protect you from a bug if you ever move these lines around and forget that the name is overloaded.

Comment (bajajneha27) [reply google-apis-core/lib/google/apis/core/storage_upload.rb]: Makes sense. Thanks for pointing it out.

**Review by frankyn — APPROVED** [review]

**Review by dazuma — APPROVED** [review]

**Pull request merged**
